[TG-3657] Always load CProver.nondetInitialize

LAJW opened the pull request:
This PR adds `CProver.nondetInitialize` to the white-list of the lazy loader.

TG Bump and test: https://github.com/diffblue/test-gen/pull/1880

Commit:
Always load cprover-nondet-initialize

smowton commented on jbmc/src/java_bytecode/ci_lazy_methods.cpp:
```diff
@@ -210,6 +210,14 @@ bool ci_lazy_methodst::operator()(
+    const irep_idt cprover_validate =
+      id2string(class_name) + ".cproverNondetInitialize:()V";
+    if(symbol_table.symbols.count(cprover_validate))
+      methods_already_populated.insert(cprover_validate);
```
Surely you should also actually populate them!

LAJW replied:
I don't have to populate anything - all of these variables are local and are used to produce this `methods_already_populated` set, which is then used to filter out things to not remove (it looks like). Most of them are badly named. For instance this one should be called `methods_to_load` and probably generated by a separate method.

I initially wanted to refactor this function, but since lazy loading V2 development is in full swing, I decided that would be a wasted effort. If my assumption was wrong, I'll create another PR that cleans this up, and leave this one alone for the sake of merging this fix on time.

smowton replied:
I meant, you should ensure the method actually is converted, not just put it in an "already_populated" set. Doing that will stop it being removed from the symbol table; putting it into the worklist will get the symbol's value (the method body) populated.

LAJW replied:
But the method body is populated. Otherwise it wouldn't fix the test-gen bug, which is what this change was all about. Really, the only problem is with preexisting misleading variable names.

smowton replied:
Any idea how it's getting populated without being added to `methods_already_populated` (whose name is accurate, it is an accumulation of methods that have been populated by the method converting loop above)?

romainbrenguier replied:
Maybe because there is a special case for it in `add_needed_class`? https://github.com/diffblue/cbmc/blob/develop/jbmc/src/java_bytecode/ci_lazy_methods_needed.cpp#L41

smowton replied:
That would result in its being added to the worklist and converted (and thus retained) in the usual way. Somehow its body is being filled in without going through the `ci_lazy_methodst` mechanism. Last I heard @LAJW was trying to find out why and submit a fix PR (or justify the status quo)

thk123 replied:
Why it works:
After the `typecheck` phase in which the `ci_lazy_methodst` is run, there is a separate `lazy_goto_model.load_all_functions();` which ensures all symbols are loaded so actually loads the method. 

This fix _shouldn't_ be required because, as @romainbrenguier points out, it is already added to the needed methods by `ci_lazy_methods_needed`. The problem (I suppose) is that a return type of an opaque method is not sufficient to cause loading that type. This is certainly wrong for test-gen, which will stub that method and try and create the object of that type. @smowton do you have an oppinion on whether that is correct for CBMC? That is:

```java
class Foo {
  public int field;
}
class Opaque {
  static Foo opaqueMethod() { return null; }
}
class EntryClass {
  static int entryFunction() { 
    return Opaque.opaqueMethod().field;
  }
}
```
Should `Foo` be loaded?

allredj commented on jbmc/src/java_bytecode/ci_lazy_methods.cpp:
```diff
@@ -210,6 +210,14 @@ bool ci_lazy_methodst::operator()(
     }
   }
 
+  for(const auto &class_name : instantiated_classes)
```
It might be useful to document the 'why' here, as `cproverNondetInitialize` doesn't mean much for jbmc itself.

romainbrenguier replied:
`cproverNondetInitialize` is meaningful for jbmc, it allows to specify invariants of the data-structures used in the program that is analysed, (for instance the length of a list is always greater than 0)

allredj reviewed (approved):
Looks alright to me and according to https://github.com/diffblue/test-gen/pull/1880 is actually fixes the bug. I would seek approval from @thk123, @smowton, or @romainbrenguier to be sure.

Commit:
Add test verifying that the cproverNondetInitialize method is always loaded

Test is disabled because of test-runner problems described in the
following issue: https://github.com/diffblue/cbmc/pull/2267

peterschrammel reviewed (commented):
The test case doesn't pass yet: https://travis-ci.org/diffblue/cbmc/jobs/387758779#L4781

LAJW commented:
@peterschrammel I've created and linked an issue in the regression test: https://github.com/diffblue/cbmc/issues/2273

peterschrammel reviewed (approved):
Let's go with that. #2273 needs to be investigated.

smowton commented:
It appears this was merged ignoring the comment above? In future I will make sure to actually block such PRs to ensure a response. In the meantime I will submit a fix of my own.

LAJW commented:
@smowton I responded in a comment, that It looks like there's nothing to fix. You dind't reply. I treated this as a neutral stance on your side.

tautschnig commented:
@LAJW Isn't https://github.com/diffblue/cbmc/pull/2267#discussion_r192749804 the reply that @smowton did send?

LAJW commented:
@tautschnig I've completely missed it in the swarm of e-mails.
